v10.0.0 tracking

rvagg opened the pull request:
Includes both #197 and #196. Will comment in both of those what I've done.

I've also applied some additional linting to get #162 to jive with #167 and have also made js-test-and-release.yml run `on: [push, pull_request]` for now, that'll get updated soon enough with https://github.com/protocol/.github/blob/master/templates/.github/workflows/js-test-and-release.yml (although I'm not sure why it's so restrictive).

Unfortunately for some reason, bringing both of the big PRs together makes aegir / pw-test fail to see the `"browser"` field for the `hashes/sha2` export and it borks on not being able to find `"crypto"`. I haven't figured out what's going on with that and I'm not sure what would have changed to start triggering that when it isn't happening in the `esm-migration` branch. You can see what's changed here: https://github.com/multiformats/js-multiformats/compare/proposal/v10.0.0..esm-migration

Open remaining items:
- [ ] https://github.com/multiformats/js-multiformats/pull/202

Items not in scope:
- [ ] https://github.com/multiformats/js-multiformats/pull/203

Items we'll be able to address afterwards as a result:
- [ ] https://github.com/protocol/.github/pull/382
- [ ] https://github.com/ipld/js-dag-pb/pull/52
- [ ] https://github.com/ipld/js-bitcoin/pull/7

rvagg commented:
Actions isn't showing the failing tests, it looks like `test:browser` isn't enough to trigger browser tests with the js-test-and-maybe-release.yml runners: https://github.com/multiformats/js-multiformats/runs/8201994927, so we'll have to address that.

rvagg commented:
pinging @achingbrain who might be able to see what's going on with the browser tests

```
✘ [ERROR] Could not resolve "crypto"

    src/hashes/sha2.js:3:19:
      3 │ import crypto from 'crypto'
        ╵                    ~~~~~~~~

  The package "crypto" wasn't found on the file system but is built into node. Are you trying to
  bundle for node? You can use "platform: 'node'" to do that, which will remove this error.

✖ Running tests failed.
Error: Build failed with 1 error:
src/hashes/sha2.js:3:19: ERROR: Could not resolve "crypto"
```

that's supposed to be addressed with the browser export for `./hashes/sha2` but now it doesn't work, even though it did in #196.

achingbrain commented:
> and have also made js-test-and-release.yml run on: [push, pull_request] for now

Any changes to this file will get overwritten by the Unified CI bot, you should PR the [source file](https://github.com/protocol/.github/blob/master/templates/.github/workflows/js-test-and-release.yml) instead.  The change looks fine though, I don't think there's any need for it to be so restrictive.

> what's going on with the browser tests

Just looking into this.  Normally you'd use the `"browser"` field in `package.json` to specify browser overrides for internal files but that doesn't seem to be specified in this project, I'm guessing because IPJS would previously have overwritten it?

I think it might be related to these sorts of imports in the tests:

```
import { sha256 } from 'multiformats/hashes/sha2'
```

Before I go any futher in investigating is there any reason the import isn't using relative paths instead of trying to resolve multiformats from `node_modules`?

RangerMauve commented:
> Just looking into this. Normally you'd use the "browser" field in package.json to specify browser overrides for internal files but that doesn't seem to be specified in this project

This project is relying on [ESM Exports](https://webpack.js.org/guides/package-exports/#target-environment) in the package.json and using the `browser` field for the target environment. What is Aegir using for the bundling before it runs tests in the browser? Does it maybe need to have it's Webpack version updated?

achingbrain commented:
> What is Aegir using for the bundling before it runs tests in the browser?

It uses [esbuild](https://github.com/ipfs/aegir/blob/master/src/build/index.js#L82-L108) for bundling.

RangerMauve commented:
Weird. Esbuild should be detecting the `exposts` field as of 0.9.0 and aegir is using a much newer one. https://github.com/evanw/esbuild/issues/187

achingbrain commented:
Yeah, the whole thing is a bit weird to be honest - this module depends on `@ipld/dag-pb` which depends on `multiformats@9.8.1`, so it gets hoisted to the `node_modules` root for me.

The test then has:

```js
import { sha256 } from 'multiformats/hashes/sha2'
```

..so it's going to resolve to `node_modules/multiformats/...` at which point the exports map will be consulted, the "browser" field found and `node_modules/multiformats/src/hashes/sha2-browser.js` returned.

But, we're importing from the installed `node_modules/multiformats` and not `../src/hashes/sha2.js` as I think is the intention here.

If it was:

```js
import { sha256 } from '../src/hashes/sha2.js'
```

with an override in the `"browser"` field (not "exports" since this is an internal import path) it should work?

RangerMauve commented:
Aha! Good digging. I think it might actually be related to the package.json in the ts-use folder trying to load multiformats from the [dist folder](https://github.com/multiformats/js-multiformats/blob/master/test/ts-use/package.json#L5) which no longer exists in this version

RangerMauve commented:
I think we need to tests to import the `ts-use` module as is rather than importing the main.ts file from it.

RangerMauve commented on tsconfig.json:
```diff
@@ -26,13 +26,29 @@
   "include": [
-    "src"
+    "src",
+    "test"
```
Was this change necessary? I think it might be what was busting the ts-use code.

RangerMauve commented:
Err, the error defs isn't due to the ts-use folder. 🤔

RangerMauve commented:
Did some digging in the build, and this is what's happening when it tries to import sha2 from the "test" directory:

<details>

```
⬥ [VERBOSE] Resolving import "multiformats/hashes/sha2" in directory "/home/mauve/programming/js-multiformats/test" of type "import-statement"

  Read 11 entries for directory "/home/mauve/programming/js-multiformats/test"
  No "browser" map found in directory "/home/mauve/programming/js-multiformats/test"
  Searching for "multiformats/hashes/sha2" in "node_modules" directories starting from "/home/mauve/programming/js-multiformats/test"
    Matching "multiformats/hashes/sha2" against "paths" in "/home/mauve/programming/js-multiformats/tsconfig.json"
      Using "/home/mauve/programming/js-multiformats" as "baseURL"
      Found a fuzzy match for "multiformats/*" in "paths"
      Attempting to load "/home/mauve/programming/js-multiformats/src/hashes/sha2.js" as a file
        Checking for file "sha2.js"
        Found file "sha2.js"
  Read 15 entries for directory "/home/mauve/programming/js-multiformats/src"
  Read 7 entries for directory "/home/mauve/programming/js-multiformats/src/hashes"
  This import is under the effect of "/home/mauve/programming/js-multiformats/tsconfig.json"
  Primary path is "/home/mauve/programming/js-multiformats/src/hashes/sha2.js" in namespace "file"

⬥ [VERBOSE] Resolving import "multiformats/hashes/sha2" in directory "/home/mauve/programming/js-multiformats/test" of type "import-statement"

  Read 11 entries for directory "/home/mauve/programming/js-multiformats/test"
  No "browser" map found in directory "/home/mauve/programming/js-multiformats/test"
  Searching for "multiformats/hashes/sha2" in "node_modules" directories starting from "/home/mauve/programming/js-multiformats/test"
    Matching "multiformats/hashes/sha2" against "paths" in "/home/mauve/programming/js-multiformats/tsconfig.json"
      Using "/home/mauve/programming/js-multiformats" as "baseURL"
      Found a fuzzy match for "multiformats/*" in "paths"
      Attempting to load "/home/mauve/programming/js-multiformats/src/hashes/sha2.js" as a file
        Checking for file "sha2.js"
        Found file "sha2.js"
  Read 7 entries for directory "/home/mauve/programming/js-multiformats/src/hashes"
  This import is under the effect of "/home/mauve/programming/js-multiformats/tsconfig.json"
  Primary path is "/home/mauve/programming/js-multiformats/src/hashes/sha2.js" in namespace "file"

```

</details>

I think the issue is that the tests are trying to import from `multiformats/` rather than `../` as @achingbrain alluded to

RangerMauve commented:
Removing the "paths" section from tsconfig has yielded the following errors in the build:

<details>

```
test/test-block.spec.js:66:37 - error TS2322: Type 'string' is not assignable to type 'CID | Record<string, any>'.

66       assert.deepStrictEqual(ret, { value: 'skip' })
                                       ~~~~~

  node_modules/multiformats/types/src/block.d.ts:101:9
    101         value: CID;
                ~~~~~
    The expected type comes from property 'value' which is declared here on type '{ value: CID; remaining: string; } | { value: Record<string, any>; remaining?: undefined; }'

test/test-cid.spec.js:243:29 - error TS2339: Property 'link' does not exist on type 'CID'.

243       assert.equal(cid, cid.link())
                                ~~~~

test/test-link.spec.js:3:23 - error TS2307: Cannot find module 'multiformats/link' or its corresponding type declarations.

3 import * as Link from 'multiformats/link'
                        ~~~~~~~~~~~~~~~~~~~

test/test-link.spec.js:73:7 - error TS2578: Unused '@ts-expect-error' directive.

73       // @ts-expect-error - types can not be inferred
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

test/test-link.spec.js:95:9 - error TS2578: Unused '@ts-expect-error' directive.

95         // @ts-expect-error - version is 1 not 0
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

</details>

RangerMauve commented:
Err, nvm I think I messed stuff up with that. 😅

RangerMauve commented:
Following from this discussion: https://github.com/ipld/js-dag-pb/pull/56#discussion_r964149849


I don't think we can have the `import from "multiformats/whatever"` style format with ESM sadly. :/ I think we have to manually specify which file we want to import from, and that also means that browser support becomes harder.

RangerMauve commented:
I think in order to get the raw specifiers thing to work, we'll need to treat the tests as a separate module similar to how ts-use does it.

BigLep commented:
2022-09-06 triage conversation: going to discuss async what the next steps are.

rvagg commented:
@RangerMauve I hope you don't mind but I'm going to force push your latest commit out of here since the tsconfig settings are to help get nicer local TS imports and they shouldn't impact external consumers.

So .. thanks for the input @achingbrain, you're right about the import picking up the one in node_modules, it's been so long since we set all this weird layout up that I'd forgotten how most of it worked. IIRC what we have is something like:

* We switched to the named imports, `multiformats/...` because that works in some bundlers (I think rollup and webpack allow this?), I think it was supposed to become a standard ESM thing (I thought it was already in Node.js but @RangerMauve corrected me today). But the main thing it let us do here was allow the rollup-based compile step to pick it all apart and put it back together again in both esm and cjs forms with appropriate relative links.
* ipjs does indeed set up the `"browser"` field, I've been hiding behind ipjs long enough that I'd forgotten this is still necessary for some bundlers and that we were simulating universal export map `"browser"` field support by automatically building it. But alas I guess we need both for some time still?
* The dag-pb import here is weird; it looks like we let it slip in as part of the traversal.js work (https://github.com/multiformats/js-multiformats/commit/2bd8878726120a69273f07e177e933734c2bf4db) but it's really not necessary because it's only forming dag-pb _shaped_ objects to traverse. So I've removed it as a dependency entirely.
* Removing dag-pb is nice, but it doesn't make multiformats go away as a transitive dependency, it comes back in now via `uint8arrays` (which uses it for the bases) through `aegir`, so it's probably going to be difficult to rid ourselves of it as a dependency unless we chase it away down there.

So my current push to this branch removes the named dependencies and uses relative ones, plus the `"browser"` field does seem to make esbuild happy.

I've also added in most of the build scripts @achingbrain suggested in https://github.com/ipld/js-dag-pb/pull/56, although I've customised the `"test"` script because (a) I'd like linting in test, (b) it complains about coverage not being supported in webworkers (even though `--cov` isn't supplied—I guess it's a default? I've had to add `--no-cov` for all but the `"test:chrome"` script to make them work) (c) the electron test inclusion borks for me because I don't have my DISPLAY setup locally like it wants (I _could_ set it up properly for it to work but I'd like to minimise hiccups like this for other contributors and electron tests can be done in CI) and (d) I wanted to get the ts-use tests in there.. Critiques of my approach?

Also, it's a bit sad to move away from named imports for tests, it was nice consuming the package from the tests according to what's exported. But that's going to be pretty hard without the ipjs toolchain doing the work for us. We could go with the ts-use style sub-package just for testing, but even then, we have `multiformats` in the node_modules, is there a possibility we'll still end up with confusion about which version it's importing _even_ if we use `"file:../../"` as the import? And, is it worth it? :shrug:

rvagg commented:
I've published a 10.0.0-pre.1 with this as it is now (FYI @Gozala in case you want to try out the Link work on something)

rvagg commented:
10.0.0-pre.2 now published with a proper pack that includes types and other pieces that were excluded by `"files"`

RangerMauve commented on package.json:
```diff
@@ -227,14 +240,49 @@
+      "bundle": false
+    },
+    "test": {
+      "cov": true,
```
```suggestion
```

Should we disable coverage being on by default and opt-into it in just the Node tests if the webworker tests aren't supporting it?

I had also experimented with not running webworker tests and just doing node and the browser. 🤔

RangerMauve commented on test/ts-use/package.json:
```diff
@@ -2,9 +2,18 @@
+    "multiformats": "file:../../"
   },
   "scripts": {
     "test": "npm install && npm_config_yes=true npx -p typescript tsc --noEmit"
```
IF this is using npx, should we pin to a specific version of typescript?

Might cause issues down the line.

rvagg replied:
yeah .. except then it needs to stay up to date, so it'd need to become a dependency in here and then have the dependabot config check this file

_or_ we could just reach back into `../node_modules/.bin/tsc` since we know it's there and the version will stay pinned to whatever the parent is using

Gozala commented on package.json:
```diff
@@ -1,159 +1,168 @@
-      "types/*": [
-        "types/*"
-      ]
+  "eslintConfig": {
```
I would really love if we tap into @hugomrdias's config skills, that are able to reconcile prettier & eslint configs.

I personally hate touching this repo because my editor tends to reformat everything with prettier which makes eslint unhappy so often have to disable prettier and then try to manually fix all the stuff.

@hugomrdias seems to have a way to configure repos such that tells both eslint & prettier how to work together in synergy.

hugomrdias replied:
these are my confs https://github.com/hugomrdias/hd-scripts ts, eslint, prettier working happily together

Gozala commented on src/bases/interface.ts:
```diff
@@ -8,9 +8,10 @@ export interface BaseEncoder {
    * @param bytes
    */
-  baseEncode(bytes: Uint8Array): string
+  baseEncode: (bytes: Uint8Array) => string
```
Nit: I really dislike this eslint rule that reformat things this way. When I see new version I interpret it as `this` does not matter, while former version tells me you should call it as `instace.baseEncode`.

rvagg replied:
not sure we have a choice if we're just wholesale importing this stuff via aegir do we? I don't feel too attached to the style here, it makes sense either way to my eyes but I do hate churn for churn's sake.

Gozala commented on src/cid.js:
```diff
@@ -156,19 +227,23 @@ export class CID {
     throw new Error('Deprecated, use .toString()')
   }
 
   get codec () {
```
Maybe this is a good opportunity to remove these deprecated fields ?

Gozala commented on src/cid.js:
```diff
@@ -156,19 +227,23 @@ export class CID {
+    return Boolean(value && (value[cidSymbol] || value.asCID === value))
   }
 
   get toBaseEncodedString () {
```
Maybe this is a good opportunity to remove these deprecated fields ?

Gozala commented on src/cid.js:
```diff
@@ -156,19 +227,23 @@ export class CID {
   }
 
   get buffer () {
-    throw new Error('Deprecated .buffer property, use .bytes to get Uint8Array instead')
```
Maybe this is a good opportunity to remove these deprecated fields ?

Gozala commented on src/cid.js:
```diff
@@ -156,19 +227,23 @@ export class CID {
+    )
   }
 
   get multibaseName () {
```
Maybe this is a good opportunity to remove these deprecated fields ?

Gozala reviewed (approved):
Looks good to me, I made few minor comments, but nothing I'd block a release on!

achingbrain commented:
> 10.0.0-pre.2 now published with a proper pack that includes types and other pieces that were excluded by "files"

Ah, in other repos we assume it'll extend the aegir tsconfig which uses `"/dist"` instead of `"/types"` - see tsconfig.json the [ipld/js-dag-pb PR](https://github.com/ipld/js-dag-pb/pull/56/files#diff-b55cdbef4907b7045f32cc5360d48d262cca5f94062e353089f189f4460039e0) for what I mean.

rvagg commented:
Some TODO's for anyone in this thread to get this released:

* [ ] Remove deprecated fields and notices (which will reduce noise from the CID tests too, yay!)
* [ ] Remove the `"cov": true` field and then get rid of the `--no-cov`'s I added
* [ ] Move published types into ./dist/ to follow the "everything generated for publish goes into ./dist/" convention we apparently have
* [ ] Something something eslint prettier - sorry but I'm not going to be messing more with this stuff but someone else is welcome to
* [ ] Rebase back on to master now we have a new commit with a 9.9.0 release

I'll try and put some work in here this week but I have at least one other thing on my plate that is trumping this in importance unfortunately.

rvagg commented:
@achingbrain when I replace most of the tsconfig.json settings with `"extends": "aegir/src/config/tsconfig.aegir.json"` I get `"emitDeclarationOnly": false` and therefore the dist/ fills up with .js files as well. Is this intended? I'm currently going with these overrides:

```json
    "outDir": "dist/types",
    "emitDeclarationOnly": true,
    "importsNotUsedAsValues": "preserve"
```

(the last one is for some funky `API` usage @Gozala has added that I don't want to undo)

That gives us types _only_ going into dist/types but nothing else in dist/. That seems to leave us with a bundle that works. Exports are coming out of src/* and their equivalent types out of dist/types/src/*, with the root `"types"` pointing to `"./dist/types/src/index.d.ts"` and the `"typesVersions"` mappings all pointing in there too.

I'm also leaving `"path"` overrides in tsconfig for the `multiformats` named imports because we use that a bit in here and that override works nicely, without reaching weirdly into node_modules/multiformats (I can be sure of this because it'd be breaking with the changes on this branch compared to ^9).

rvagg commented:
`multiformats@10.0.0-pre.3` published if anyone wants to poke around in the bundle, the only thing not done in the above list is the eslint/prettier things.

RangerMauve commented:
Cool, I'll try importing the prerelease and messing with it in ts and ESM.

RangerMauve commented:
Works fine in JS with ESM, trying to import stuff from Typescript caused some errors though:

```typescript
import { CID } from 'multiformats/cid'
import * as json from 'multiformats/codecs/json'
import { sha256 } from 'multiformats/hashes/sha2'

const bytes = json.encode({ hello: 'world' })

const hash = await sha256.digest(bytes)
const cid = CID.create(1, json.code, hash)
console.log(cid)
//> CID(bagaaierasords4njcts6vs7qvdjfcvgnume4hqohf65zsfguprqphs3icwea)
```

tsconfig.json: 

```json
{
  "compilerOptions": {
    "module": "nodenext"
  },
  "files": [
    "example.ts"
  ]
}
```

<details>
<summary>
Error log
</summary>

```
node_modules/multiformats/dist/types/src/codecs/json.d.ts(3,44): error TS2834: Relative import paths need explicit file extensions in EcmaScript imports when '--moduleResolution' is 'node16' or 'nodenext'. Consider adding an extension to the import path.
node_modules/multiformats/dist/types/src/codecs/json.d.ts(4,40): error TS2834: Relative import paths need explicit file extensions in EcmaScript imports when '--moduleResolution' is 'node16' or 'nodenext'. Consider adding an extension to the import path.
node_modules/multiformats/dist/types/src/codecs/json.d.ts(5,34): error TS2834: Relative import paths need explicit file extensions in EcmaScript imports when '--moduleResolution' is 'node16' or 'nodenext'. Consider adding an extension to the import path.
node_modules/multiformats/dist/types/src/hashes/digest.d.ts(3,69): error TS2834: Relative import paths need explicit file extensions in EcmaScript imports when '--moduleResolution' is 'node16' or 'nodenext'. Consider adding an extension to the import path.
node_modules/multiformats/dist/types/src/hashes/digest.d.ts(31,38): error TS2834: Relative import paths need explicit file extensions in EcmaScript imports when '--moduleResolution' is 'node16' or 'nodenext'. Consider adding an extension to the import path.
node_modules/multiformats/dist/types/src/hashes/hasher.d.ts(32,58): error TS2834: Relative import paths need explicit file extensions in EcmaScript imports when '--moduleResolution' is 'node16' or 'nodenext'. Consider adding an extension to the import path.
node_modules/multiformats/dist/types/src/link/interface.d.ts(1,38): error TS2834: Relative import paths need explicit file extensions in EcmaScript imports when '--moduleResolution' is 'node16' or 'nodenext'. Consider adding an extension to the import path.
node_modules/multiformats/dist/types/src/link/interface.d.ts(2,68): error TS2834: Relative import paths need explicit file extensions in EcmaScript imports when '--moduleResolution' is 'node16' or 'nodenext'. Consider adding an extension to the import path.
node_modules/multiformats/dist/types/src/link/interface.d.ts(3,40): error TS2834: Relative import paths need explicit file extensions in EcmaScript imports when '--moduleResolution' is 'node16' or 'nodenext'. Consider adding an extension to the import path.
```

</details>

Looks like it's expecting extensions on imports since the "exports" field isn't being considered?

rvagg commented:
great, thanks @RangerMauve! I've replicated it in ts-use and applied some fixes that appear to take care of this.

@Gozala would you mind sanity checking 04392d2 for me on this please?

10.0.0-pre.4 published.

Gozala commented:
> @Gozala would you mind sanity checking [04392d2](https://github.com/multiformats/js-multiformats/commit/04392d2c2330283e3d3e84f82238df5f166bf387) for me on this please?

Looks good to me! Thanks @rvagg

achingbrain commented:
> @achingbrain when I replace most of the tsconfig.json settings with "extends": "aegir/src/config/tsconfig.aegir.json" I get "emitDeclarationOnly": false and therefore the dist/ fills up with .js files as well. Is this intended?

Yes, the default position for aegir's `tsconfig.json` is for TS rather than ESM so all the ESM+types modules extend and [override `emitDeclarationOnly`](https://github.com/ipfs/js-ipfs-bitswap/blob/master/tsconfig.json#L5) in the same way as here.

achingbrain commented on package.json:
```diff
@@ -227,14 +240,48 @@
+      ]
+    }
+  },
+  "main": "./src/index.js"
```
Could probably lose the `"main"` field since `"exports"` is specified and we're all future-tech now?

rvagg commented:
I think we're good to go here. I'll remove the `"main"` from package.json as suggested above and then publish a 10.0.0, tomorrow hopefully.

rvagg commented:
Let's get #202 in first to join the breakage party.

rvagg commented:
blocked by critique in #203

achingbrain commented:
I don't think #203 should block this, it can be resolved later.

rvagg commented:
* pulled in #202 with changes (see that thread)
* rebased to current master
* removed `"main"`

Ready to roll, 🤞 hoping that auto-release will give us a proper v10.0.0 on merge.

Any final comments or objections?

achingbrain commented:
> hoping that auto-release will give us a proper v10.0.0 on merge

It should do as long as the commit title has a `!` (e.g. `feat!: ...`) and/or `BREAKING CHANGE: ...` is in the message.

BigLep commented:
@rvagg : I think it's time to pull the trigger :)

rvagg commented:
@BigLep yeah, I ran out of time yesterday to get the auto-release process tested - that it's not going to have problems with my pre-releases and that it's going to pick up a proper semver-major release for this. Thankfully I did bother testing because it wasn't going to semver-major this, I'd assumed that one of us had tagged at least one commit as breaking, but nope!

Manually merging and manipulating the commit history for a bit of squashing and rewriting of comments. Will push that to master and it should release a 10.0.0.
